File: utils/com_segmentation.py
"""
com_segmentation.py — Segment-weighted whole-body COM in 2D normalized image space.

Shared by ``feature_extractor`` (per-frame series for the API) and ``visualization``
(annotated video). Logic mirrors the biomechanics literature (mass fractions +
COM along segment chords); see module docstring in the thesis / API docs.
"""

from __future__ import annotations

import numpy as np

# Mass fractions of total body mass (Plagenhoef / Winter–style anthropometry).
_SEGMENT_MASS = {
    "head": 8.1,
    "trunk": 46.7,
    "l_upper_arm": 2.8,
    "r_upper_arm": 2.8,
    "l_forearm": 1.6,
    "r_forearm": 1.6,
    "l_hand": 0.6,
    "r_hand": 0.6,
    "l_thigh": 10.0,
    "r_thigh": 10.0,
    "l_shank": 4.65,
    "r_shank": 4.65,
    "l_foot": 1.43,
    "r_foot": 1.43,
}
_FRAC_TRUNK = 0.43
_FRAC_UPPER_ARM = 0.436
_FRAC_FOREARM = 0.430
_FRAC_HAND = 0.50
_FRAC_THIGH = 0.433
_FRAC_SHANK = 0.433
_FRAC_FOOT = 0.50
_VIS_FLOOR = 0.18


def _vis(lm: np.ndarray, i: int) -> float:
    row = lm[i]
    if len(row) > 3:
        return float(max(0.0, min(1.0, row[3])))
    return 1.0


def _seg_weight(lm: np.ndarray, *idx: int) -> float:
    v = min(_vis(lm, i) for i in idx)
    return max(0.0, v)


def _com_on_segment(lm: np.ndarray, a: int, b: int, frac: float) -> np.ndarray:
    pa = lm[a][:2].astype(np.float64)
    pb = lm[b][:2].astype(np.float64)
    return pa + frac * (pb - pa)


def compute_segment_weighted_com_xy(landmarks: list | np.ndarray) -> tuple[float, float] | None:
    """
    Approximate whole-body COM in **normalized image coordinates** (same as landmarks).

    Returns ``(x, y)`` in [0, 1] normalized space, or ``None`` if nothing reliable remains.
    """
    lm = np.asarray(landmarks, dtype=np.float64)
    if lm.shape[0] < 33:
        return None

    ear_mid = (lm[7][:2] + lm[8][:2]) / 2.0
    head_com = (lm[0][:2] + ear_mid) / 2.0
    w_head = _seg_weight(lm, 0, 7, 8) * _SEGMENT_MASS["head"]

    hip_mid = (lm[23][:2] + lm[24][:2]) / 2.0
    shoulder_mid = (lm[11][:2] + lm[12][:2]) / 2.0
    trunk_vec = shoulder_mid - hip_mid
    trunk_com = hip_mid + _FRAC_TRUNK * trunk_vec
    w_trunk = _seg_weight(lm, 11, 12, 23, 24) * _SEGMENT_MASS["trunk"]

    segments: list[tuple[float, np.ndarray]] = [
        (w_head, head_com),
        (w_trunk, trunk_com),
        (_seg_weight(lm, 11, 13) * _SEGMENT_MASS["l_upper_arm"], _com_on_segment(lm, 11, 13, _FRAC_UPPER_ARM)),
        (_seg_weight(lm, 12, 14) * _SEGMENT_MASS["r_upper_arm"], _com_on_segment(lm, 12, 14, _FRAC_UPPER_ARM)),
        (_seg_weight(lm, 13, 15) * _SEGMENT_MASS["l_forearm"], _com_on_segment(lm, 13, 15, _FRAC_FOREARM)),
        (_seg_weight(lm, 14, 16) * _SEGMENT_MASS["r_forearm"], _com_on_segment(lm, 14, 16, _FRAC_FOREARM)),
        (_seg_weight(lm, 15, 19) * _SEGMENT_MASS["l_hand"], _com_on_segment(lm, 15, 19, _FRAC_HAND)),
        (_seg_weight(lm, 16, 20) * _SEGMENT_MASS["r_hand"], _com_on_segment(lm, 16, 20, _FRAC_HAND)),
        (_seg_weight(lm, 23, 25) * _SEGMENT_MASS["l_thigh"], _com_on_segment(lm, 23, 25, _FRAC_THIGH)),
        (_seg_weight(lm, 24, 26) * _SEGMENT_MASS["r_thigh"], _com_on_segment(lm, 24, 26, _FRAC_THIGH)),
        (_seg_weight(lm, 25, 27) * _SEGMENT_MASS["l_shank"], _com_on_segment(lm, 25, 27, _FRAC_SHANK)),
        (_seg_weight(lm, 26, 28) * _SEGMENT_MASS["r_shank"], _com_on_segment(lm, 26, 28, _FRAC_SHANK)),
        (_seg_weight(lm, 27, 31) * _SEGMENT_MASS["l_foot"], _com_on_segment(lm, 27, 31, _FRAC_FOOT)),
        (_seg_weight(lm, 28, 32) * _SEGMENT_MASS["r_foot"], _com_on_segment(lm, 28, 32, _FRAC_FOOT)),
    ]

    weighted: list[tuple[float, np.ndarray]] = []
    for mass, pt in segments:
        if mass < _VIS_FLOOR:
            continue
        weighted.append((mass, pt))

    if not weighted:
        return None

    total_w = sum(m for m, _ in weighted)
    if total_w < 1e-8:
        return None

    com = sum(m * p for m, p in weighted) / total_w
    x, y = float(com[0]), float(com[1])
    if not np.isfinite(x) or not np.isfinite(y):
        return None
    return (x, y)

File: frontend/src/utils/comSegmentation.ts
/**
 * Segment-weighted whole-body COM in normalized image coordinates.
 * Mirrors `utils/com_segmentation.py` for client-side use with `/poses/{id}` JSON.
 */

import type { ComXY, PosesJson } from "../api/types";

const SEGMENT_MASS = {
  head: 8.1,
  trunk: 46.7,
  l_upper_arm: 2.8,
  r_upper_arm: 2.8,
  l_forearm: 1.6,
  r_forearm: 1.6,
  l_hand: 0.6,
  r_hand: 0.6,
  l_thigh: 10.0,
  r_thigh: 10.0,
  l_shank: 4.65,
  r_shank: 4.65,
  l_foot: 1.43,
  r_foot: 1.43,
} as const;

const FRAC_TRUNK = 0.43;
const FRAC_UPPER_ARM = 0.436;
const FRAC_FOREARM = 0.43;
const FRAC_HAND = 0.5;
const FRAC_THIGH = 0.433;
const FRAC_SHANK = 0.433;
const FRAC_FOOT = 0.5;
const VIS_FLOOR = 0.18;

function vis(lm: number[][], i: number): number {
  const row = lm[i];
  if (row.length > 3) return Math.max(0, Math.min(1, row[3]!));
  return 1;
}

function segWeight(lm: number[][], idx: number[]): number {
  let v = 1;
  for (const i of idx) {
    v = Math.min(v, vis(lm, i));
  }
  return Math.max(0, v);
}

function comOnSegment(lm: number[][], a: number, b: number, frac: number): [number, number] {
  const pax = lm[a]![0]!;
  const pay = lm[a]![1]!;
  const pbx = lm[b]![0]!;
  const pby = lm[b]![1]!;
  return [pax + frac * (pbx - pax), pay + frac * (pby - pay)];
}

/**
 * Approximate whole-body COM in normalized [0,1] image coordinates.
 */
export function computeSegmentWeightedComXY(landmarks: number[][]): ComXY | null {
  if (landmarks.length < 33) return null;

  const lm = landmarks;

  const earMidX = (lm[7]![0]! + lm[8]![0]!) / 2;
  const earMidY = (lm[7]![1]! + lm[8]![1]!) / 2;
  const headComX = (lm[0]![0]! + earMidX) / 2;
  const headComY = (lm[0]![1]! + earMidY) / 2;
  const wHead = segWeight(lm, [0, 7, 8]) * SEGMENT_MASS.head;

  const hipMidX = (lm[23]![0]! + lm[24]![0]!) / 2;
  const hipMidY = (lm[23]![1]! + lm[24]![1]!) / 2;
  const shoulderMidX = (lm[11]![0]! + lm[12]![0]!) / 2;
  const shoulderMidY = (lm[11]![1]! + lm[12]![1]!) / 2;
  const trunkVecX = shoulderMidX - hipMidX;
  const trunkVecY = shoulderMidY - hipMidY;
  const trunkComX = hipMidX + FRAC_TRUNK * trunkVecX;
  const trunkComY = hipMidY + FRAC_TRUNK * trunkVecY;
  const wTrunk = segWeight(lm, [11, 12, 23, 24]) * SEGMENT_MASS.trunk;

  const segments: [number, [number, number]][] = [
    [wHead, [headComX, headComY]],
    [wTrunk, [trunkComX, trunkComY]],
    [segWeight(lm, [11, 13]) * SEGMENT_MASS.l_upper_arm, comOnSegment(lm, 11, 13, FRAC_UPPER_ARM)],
    [segWeight(lm, [12, 14]) * SEGMENT_MASS.r_upper_arm, comOnSegment(lm, 12, 14, FRAC_UPPER_ARM)],
    [segWeight(lm, [13, 15]) * SEGMENT_MASS.l_forearm, comOnSegment(lm, 13, 15, FRAC_FOREARM)],
    [segWeight(lm, [14, 16]) * SEGMENT_MASS.r_forearm, comOnSegment(lm, 14, 16, FRAC_FOREARM)],
    [segWeight(lm, [15, 19]) * SEGMENT_MASS.l_hand, comOnSegment(lm, 15, 19, FRAC_HAND)],
    [segWeight(lm, [16, 20]) * SEGMENT_MASS.r_hand, comOnSegment(lm, 16, 20, FRAC_HAND)],
    [segWeight(lm, [23, 25]) * SEGMENT_MASS.l_thigh, comOnSegment(lm, 23, 25, FRAC_THIGH)],
    [segWeight(lm, [24, 26]) * SEGMENT_MASS.r_thigh, comOnSegment(lm, 24, 26, FRAC_THIGH)],
    [segWeight(lm, [25, 27]) * SEGMENT_MASS.l_shank, comOnSegment(lm, 25, 27, FRAC_SHANK)],
    [segWeight(lm, [26, 28]) * SEGMENT_MASS.r_shank, comOnSegment(lm, 26, 28, FRAC_SHANK)],
    [segWeight(lm, [27, 31]) * SEGMENT_MASS.l_foot, comOnSegment(lm, 27, 31, FRAC_FOOT)],
    [segWeight(lm, [28, 32]) * SEGMENT_MASS.r_foot, comOnSegment(lm, 28, 32, FRAC_FOOT)],
  ];

  const weighted: [number, [number, number]][] = [];
  for (const [mass, pt] of segments) {
    if (mass < VIS_FLOOR) continue;
    weighted.push([mass, pt]);
  }
  if (weighted.length === 0) return null;

  let totalW = 0;
  let sx = 0;
  let sy = 0;
  for (const [m, [px, py]] of weighted) {
    totalW += m;
    sx += m * px;
    sy += m * py;
  }
  if (totalW < 1e-8) return null;

  const x = sx / totalW;
  const y = sy / totalW;
  if (!Number.isFinite(x) || !Number.isFinite(y)) return null;
  return { x, y };
}

/** Build a per-frame COM series from raw pose JSON (uncertainties differ from API-smoothed series). */
export function computeComSeriesFromPosesJson(data: PosesJson): (ComXY | null)[] {
  return data.poses.map((p) => {
    if (!p.landmarks || p.landmarks.length < 33) return null;
    return computeSegmentWeightedComXY(p.landmarks);
  });
}
